src/tmpfiles/tmpfiles.c

```diff
@@ -729,32 +768,49 @@ static int dir_cleanup(
                                 log_debug("File \"%s\": modify time %s is too new.",
                                           sub_path,
                                           format_timestamp_style(a, sizeof(a), mtime_nsec / NSEC_PER_USEC, TIMESTAMP_US));
-                                continue;
-                        }
+
+                                if (age_by == AGE_BY_DEFAULT)
+                                        continue;
+                        } else
+                                subpath_age_by |= AGE_BY_MTIME_F;
```

Comment:
i don't grok this part btw? what's your intention here?

Reply:
As mentioned in the comment for `mtime`. If it is the "default" mode - skip checking for any other timestamp as soon as we find the first "too new" file timestamp type. For other cases, where the arguments are provided by the user, `subpath_age_by` keeps track of the file timestamps types for which a path has "aged by", and if they match with whatever the user has provided, proceed to delete it. Otherwise, move on to the next entry.

Reply:
Just following up, are you suggesting I should check `mtime_nsec < cutoff_nsec` in the `else` (because of `NSEC_INFINITY`)?

Reply:
i don't understand what you are trying to say? 

with the implied defaults we basically respect the newest of the timestamps. but if the values are explicitly configured you want to respect the "oldest" instead? or what are you trying to say? i don't get it?

can you maybe given an example for a file and how it should or should not be aged?

Reply:
i still don't follow. why do you do the mtime check at all if it's not in the bitmask of what to check?

Reply:
Ah, OK! I didn't follow which part you were referring to, earlier. So, yes, we can enclose each of the block like so:
```c
if ((age_by_file & AGE_BY_MTIME)) {
	if (mtime_nsec != NSEC_INFINITY && mtime_nsec >= cutoff_nsec) {
		char a[FORMAT_TIMESTAMP_MAX];
		log_debug("File \"%s\": modify time %s is too new.",
			  sub_path,
			  format_timestamp_style(a, sizeof(a), mtime_nsec / NSEC_PER_USEC, TIMESTAMP_US));

		if (age_by_file == AGE_BY_DEFAULT_FILE)
			continue;
	} else if (mtime_nsec < cutoff_nsec)
		subpath_age_by |= AGE_BY_MTIME;
}
```

I feel we should still keep the default mode check, though. For example, if we have a file with `atime` (1h in the past), and `mtime` (0.5h in the past) and the argument is `am:1h`, we first check `mtime`, and without the check enclosing the `continue`, we would never check the subsequent `atime` block.

About your previous question:

> with the implied defaults we basically respect the newest of the timestamps. but if the values are explicitly configured you want to respect the "oldest" instead? or what are you trying to say? i don't get it?

Yes. I would say specifying `amcb:1h` is not the same as `1h` (with `AGE_BY_DEFAULT_FILE`). In the former, we don't delete a file if at least one of the specified timestamp types is older than the cut-off time, whereas in the latter, we don't delete a file even if one of the timestamp types is newer than the cutoff time (this is the reason for going with `AGE_BY_UNSET` for the "default" mode, originally). So, `amcb` is like an `or`, and `AGE_BY_DEFAULT_FILE` is like an `and`.

Is the described behavior acceptable? Or, should I skip the default mode check and `continue` if even one of the timestamp types is newer than the cut-off time?

Reply:
The last paragraph from my previous comment is not clear. I just realized that when the user specifies `amcb`, it becomes `AGE_BY_DEFAULT_FILE`. I meant to say, we should not have `AGE_BY_DEFAULT_{FILE,DIR}`, but have `AGE_BY_UNSET_{FILE,DIR}` (which is set to `0` by default).

Reply:
I'm hoping that this will provide a better explanation (sorry for the noise, my previous comments were made in a hurry):

- If we decide go with checking for only the timestamps that the user wants, and even if one of them is newer than the cut-off time, skip checking for any other timestamp types (specified by the user), it would look like this:
```c
if (age_by_dir & AGE_BY_MTIME) {
	if (mtime_nsec != NSEC_INFINITY && mtime_nsec >= cutoff_nsec) {
		/* Skip deletion; don't check for other times. */
		continue;
}

/* Check for other timestamps. */
```

We would need `AGE_BY_DEFAULT_{FILE,DIR}` here, because that it has all of the `AGE_BY_*TIME` flags.


- If we go with the other option, where we check for all the timestamps specified by the user, and skip deletion when none of timestamps are newer that the cut-off time.
```c
if ((age_by_dir & AGE_BY_MTIME) || age_by_dir == AGE_BY_UNSET) {
	if (mtime_nsec != NSEC_INFINITY && mtime_nsec >= cutoff_nsec) {
		/*
		 * Default behavior: skip deletion as soon newer file
		 * timestamp is found; don't check others.
		 */
		if (age_by == AGE_BY_UNSET)
			continue;
	} else if (mtime_nsec < cutoff_nsec)
		/* Store it for checking later. */
		subpath_age_by |= AGE_BY_MTIME;
}

/* Check for other timestamps. */


/* If the file hasn't aged by any of the specified timestamp types, skip. */
if (!(subpath_age_by & age_by_dir))
	continue;
```

We would need `AGE_BY_UNSET` ( `= 0`) here, because by default (i.e., current behavior) where we skip deletion if first timestamp we check is newer than the specified age and skip checking any other subsequent timestamp types.